syntax = "proto2";

import "development/vndk/tools/header-checker/proto/abi_dump.proto";

package abi_diff;

message RecordFieldDeclDiff {
  optional abi_dump.RecordFieldDecl old = 1;
  optional abi_dump.RecordFieldDecl new = 2;
  required uint32 index = 3;
}

message EnumFieldDeclDiff {
  optional abi_dump.EnumFieldDecl old = 1;
  optional abi_dump.EnumFieldDecl new = 2;
  required uint32 index = 3;
}

message CXXBaseSpecifierDiff {
  optional abi_dump.CXXBaseSpecifier old = 1;
  optional abi_dump.CXXBaseSpecifier new = 2;
  required uint32 index = 3;
}

message CXXVTableDiff {
  optional abi_dump.VTableComponent old = 1;
  optional abi_dump.VTableComponent new = 2;
  required uint32 index = 3;
}

message BasicNamedAndTypedDeclDiff {
  required abi_dump.BasicNamedAndTypedDecl old = 1;
  required abi_dump.BasicNamedAndTypedDecl new = 2;
}

message RecordDeclDiff {
  repeated RecordFieldDeclDiff field_diffs = 1;
  repeated CXXBaseSpecifierDiff base_diffs = 2;
  repeated CXXVTableDiff vtable_diffs = 3;
  optional BasicNamedAndTypedDeclDiff type_diff = 4;
  required string name = 5;
}

message EnumDeclDiff {
  repeated EnumFieldDeclDiff field_diffs = 1;
  optional BasicNamedAndTypedDeclDiff type_diff = 2;
  required string name = 3;
}

message ReturnTypeDiff {
  required abi_dump.BasicNamedAndTypedDecl old = 1;
  required abi_dump.BasicNamedAndTypedDecl new = 2;
}

message ParamDeclDiff {
  optional abi_dump.ParamDecl old = 1;
  optional abi_dump.ParamDecl new = 2;
  required uint32 index = 3;
}

message FunctionDeclDiff {
  required ReturnTypeDiff return_type_diffs = 1;
  repeated ParamDeclDiff param_diffs = 2;
}

message GlobalVarDeclDiff {
  optional BasicNamedAndTypedDeclDiff type_diff = 1;
}

message TranslationUnitDiff {
  // Differing Elements.
  repeated RecordDeclDiff records_diff = 1;
  repeated EnumDeclDiff enums_diff = 2;
  repeated FunctionDeclDiff functions_diff = 3;
  repeated GlobalVarDeclDiff global_vars_diff = 4;
  // Removed Elements.
  repeated abi_dump.RecordDecl records_removed = 5;
  repeated abi_dump.FunctionDecl functions_removed = 6;
  repeated abi_dump.EnumDecl enums_removed = 7;
  repeated abi_dump.GlobalVarDecl global_vars_removed = 8;
  // Added Elements.
  repeated abi_dump.RecordDecl records_added = 9;
  repeated abi_dump.FunctionDecl functions_added = 10;
  repeated abi_dump.EnumDecl enums_added = 11;
  repeated abi_dump.GlobalVarDecl global_vars_added = 12;
}
